File: AccountMicroservice/src/main/java/com/accountMicroservice/service/AccountService.java
package com.accountMicroservice.service;

import java.util.List;
import java.util.Optional;

import com.accountMicroservice.dto.AccountCreationRequest;
import com.accountMicroservice.dto.AccountResponse;
import com.accountMicroservice.dto.AccountUpdateRequest;
import com.accountMicroservice.dto.DepositRequest;
import com.accountMicroservice.dto.WithdrawRequest;
import com.accountMicroservice.exception.AccountCreationException;
import com.accountMicroservice.exception.AccountNotFoundException;
import com.accountMicroservice.exception.AccountProcessingException;
import com.accountMicroservice.exception.InsufficientFundsException;

/**
 * Interface for the Account Service, defining the core business operations
 * related to managing user accounts.
 */
public interface AccountService {

    /**
     * Creates a new bank account for a user.
     * @param request The AccountCreationRequest DTO.
     * @return The created AccountResponse DTO.
     * @throws AccountCreationException if account creation fails.
     */
    AccountResponse createAccount(AccountCreationRequest request);

    /**
     * Retrieves account details by account ID.
     * @param accountId The ID of the account.
     * @return An Optional containing the AccountResponse DTO if found, or empty otherwise.
     */
    Optional<AccountResponse> getAccountById(String accountId);

    /**
     * Retrieves account details by account number.
     * @param accountNumber The account number.
     * @return An Optional containing the AccountResponse DTO if found, or empty otherwise.
     */
    Optional<AccountResponse> getAccountByAccountNumber(String accountNumber); // <--- NEW METHOD

    /**
     * Retrieves all accounts associated with a specific user ID.
     * @param userId The ID of the user.
     * @return A list of AccountResponse DTOs.
     */
    List<AccountResponse> getAccountsByUserId(String userId);

    /**
     * Updates the status of an account (e.g., ACTIVE to CLOSED).
     * @param accountId The ID of the account to update.
     * @param request The AccountUpdateRequest DTO containing the new status.
     * @return The updated AccountResponse DTO.
     * @throws AccountNotFoundException if the account is not found.
     * @throws AccountProcessingException if the update fails.
     */
    AccountResponse updateAccountStatus(String accountId, AccountUpdateRequest request);

    /**
     * Deposits funds into a specified account.
     * @param accountId The ID of the account to deposit into.
     * @param request The DepositRequest DTO.
     * @return The updated AccountResponse DTO.
     * @throws AccountNotFoundException if the account is not found.
     * @throws AccountProcessingException if the deposit fails.
     */
    AccountResponse depositFunds(String accountId, DepositRequest request);

    /**
     * Withdraws funds from a specified account.
     * @param accountId The ID of the account to withdraw from.
     * @param request The WithdrawRequest DTO.
     * @return The updated AccountResponse DTO.
     * @throws AccountNotFoundException if the account is not found.
     * @throws InsufficientFundsException if the account has insufficient funds.
     * @throws AccountProcessingException if the withdrawal fails.
     */
    AccountResponse withdrawFunds(String accountId, WithdrawRequest request);

    /**
     * Deletes or closes an account.
     * @param accountId The ID of the account to delete.
     * @throws AccountNotFoundException if the account is not found.
     * @throws AccountProcessingException if the deletion fails.
     */
    void deleteAccount(String accountId);
}

File: Angular Frontend/banking-frontend/src/app/features/accounts/account.service.ts
// src/app/features/accounts/account.service.ts

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AccountCreationRequest, AccountResponse, AccountStatus, AccountUpdateRequest, DepositRequest, WithdrawRequest } from '../../shared/models/account.model';




@Injectable({
  providedIn: 'root'
})
export class AccountService {

  private accountsApiUrl = `${environment.apiUrl}/accounts`; // Base URL for Account Microservice

  constructor(private http: HttpClient) { }

  /**
   * Creates a new bank account.
   * POST /accounts/create
   * @param request The account creation request payload.
   * @returns An Observable of the created AccountResponse.
   */
  createAccount(request: AccountCreationRequest): Observable<AccountResponse> {
    return this.http.post<AccountResponse>(`${this.accountsApiUrl}/create`, request);
  }

  /**
   * Retrieves all accounts associated with a specific user ID.
   * GET /accounts/user/{userId}
   * @param userId The ID of the user.
   * @returns An Observable of a list of AccountResponse.
   */
  getAccountsByUserId(userId: string): Observable<AccountResponse[]> {
    return this.http.get<AccountResponse[]>(`${this.accountsApiUrl}/user/${userId}`);
  }

  /**
   * Retrieves account details by account ID.
   * GET /accounts/{accountId}
   * @param accountId The ID of the account.
   * @returns An Observable of the AccountResponse.
   */
  getAccountById(accountId: string): Observable<AccountResponse> {
    return this.http.get<AccountResponse>(`${this.accountsApiUrl}/${accountId}`);
  }

  /**
   * Updates the status of an account (e.g., ACTIVE to CLOSED).
   * PUT /accounts/{accountId}
   * @param accountId The ID of the account to update.
   * @param status The new status (e.g., AccountStatus.CLOSED).
   * @returns An Observable of the updated AccountResponse.
   */
  updateAccountStatus(accountId: string, status: AccountStatus): Observable<AccountResponse> {
    const request: AccountUpdateRequest = { status: status };
    return this.http.put<AccountResponse>(`${this.accountsApiUrl}/${accountId}`, request);
  }

  /**
   * Deletes or closes an account.
   * DELETE /accounts/{accountId}
   * @param accountId The ID of the account to delete.
   * @returns An Observable of void.
   */
  deleteAccount(accountId: string): Observable<void> {
    return this.http.delete<void>(`${this.accountsApiUrl}/${accountId}`);
  }

  /**
   * Deposits funds into a specified account (internal endpoint for backend services).
   * POST /accounts/{accountId}/deposit
   * @param accountId The ID of the account.
   * @param request The deposit request payload (transactionId, amount).
   * @returns An Observable of the updated AccountResponse.
   */
  depositFunds(accountId: string, request: DepositRequest): Observable<AccountResponse> {
    return this.http.post<AccountResponse>(`${this.accountsApiUrl}/${accountId}/deposit`, request);
  }

  /**
   * Withdraws funds from a specified account (internal endpoint for backend services).
   * POST /accounts/{accountId}/withdraw
   * @param accountId The ID of the account.
   * @param request The withdrawal request payload (transactionId, amount).
   * @returns An Observable of the updated AccountResponse.
   */
  withdrawFunds(accountId: string, request: WithdrawRequest): Observable<AccountResponse> {
    return this.http.post<AccountResponse>(`${this.accountsApiUrl}/${accountId}/withdraw`, request);
  }
}